### PR TITLE
Clarify type-class policy

### PR DESCRIPTION
**Pull request by phadej**

### CODE REVIEW

File: README.markdown [hunk 1]
```diff
@@ -186,8 +186,33 @@ instance FiniteBits Foo where
...
+displayException :: Exception e => e -> String
+displayException = show
+#endif
+```
+
+As with previous approach, you won't be able to define new members of the type
+class without CPP guards. In other words, non-CPP approach would limit
+uses to lowest common denominator.
+
+As neither approach is a very unsatisfactory solution, and to embrace
```
Comment (RyanGlScott) [inline README.markdown]: I think you meant to say "neither approach is a very _satisfactory_ solution"?

File: README.markdown [hunk 2]
```diff
@@ -186,8 +186,33 @@ instance FiniteBits Foo where
...
+but have grown their API. For example, it's tempting to define
+
+```haskell
+#if !(MIN_VERSION_base(4,8,0))
+displayException :: Exception e => e -> String
+displayException = show
+#endif
+```
+
+As with previous approach, you won't be able to define new members of the type
```
Comment (RyanGlScott) [inline README.markdown]: As with _the_ previous approach

File: README.markdown [hunk 3]
```diff
@@ -186,8 +186,33 @@ instance FiniteBits Foo where
...
+class without CPP guards. In other words, non-CPP approach would limit
```
Comment (RyanGlScott) [inline README.markdown]: In other words, _the_ non-CPP approach would limit uses to _the_ lowest common denominator.

File: README.markdown [hunk 4]
```diff
@@ -186,8 +186,33 @@ instance FiniteBits Foo where
...
+additional methods outside of the class:
+
+```
+#if MIN_VERSION_base(4,7,0) && !(MIN_VERSION_base(4,8,0))
+countLeadingZeros :: FiniteBits b => b -> Int
+countLeadingZeros = {- default implementation #-}
+#endif
+```
+
+The situation is only sligtly better for classes which exist across all `base` version,
```
Comment (RyanGlScott) [inline README.markdown]: for classes which exist across all versions of `base`

(Also there's a typo: s/sligtly/slightly/)

File: README.markdown [hunk 5]
```diff
@@ -186,8 +186,33 @@ instance FiniteBits Foo where
 #endif
 ```
 
-This is a very unsatisfactory solution, and for this reason, we do not pursue
-it. For similar reasons, we do not backport data types.
+Another approach is to backport the API from `base-4.7.0.0` and to declare
+additional methods outside of the class:
+
+```
```
Comment (RyanGlScott) [inline README.markdown]:
~~~
```haskell
~~~

**Comment by sol** [comment]
I think the main reason we don't do type classes, data types instances is that some of it is already covered by other packages and redefining the same things multiple times is controversial.  So back then this was a pragmatic decision.

In general, I would still love to have one package that does as much as possible in terms of backwards compatibility and if e.g. @ekmett would agree to collaborate it would be nice to migrate stuff over to `base-compat` or maybe better `base-orphans`.

**Comment by RyanGlScott** [comment]
By the way, I fixed the GHC 7.0 failures on Travis recently on `master`, so rebasing on top of that should help.

**Comment by ekmett** [comment]
@sol: @ryanglscott has been dealing with almost all of the compatibility issues I've had for the last couple of years. If you can convince him that we should do some big sweeping move then I'd be pretty easy to convince. 

I currently use `base-orphans` almost everywhere, but I don't believe I currently have a `base-compat` dependency anywhere in my code.

**Comment by phadej** [comment]
Sorry it took long, but corrected mistakes which @RyanGlScott pointed out.

**Comment by RyanGlScott** [comment]
LGTM, thanks!

**Pull request merged**